File: 10-28-20/fib.swift
func fib(n: Int) -> Int {
  var newMemo:[Int:Int] = [:]
  return _fib(n: n, memo: &newMemo)
}

func _fib(n: Int, memo: inout Dictionary<Int, Int>) -> Int {
  guard memo[n] == nil else { return memo[n]! }
  guard n > 2 else { return 1 }

  memo[n] = _fib(n: n - 1, memo: &memo) + _fib(n: n - 2, memo: &memo)
  return memo[n]!
}

print(fib(n: 6)) // 8
print(fib(n: 7)) // 13
print(fib(n: 8)) // 21
print(fib(n: 9)) // 34
print(fib(n: 42)) // 267914296

File: 10-28-20/fib.rs
use std::collections::HashMap;

fn fib(n: u8) -> u32 {
  if n == 1 || n == 2 {
    return 1;
  }

  fib(n - 1) + fib(n - 2)
}

fn memoized_fib(n: u8, mut memo: &mut HashMap<u8, u128>) -> u128 {
  if memo.contains_key(&n) {
    return memo[&n]
  }

  if n == 1 || n == 2 {
    return 1;
  }

  let fib_num = memoized_fib(n - 1, &mut memo) + memoized_fib(n - 2, &mut memo);
  memo.insert(n, fib_num);
  memo[&n]
}

fn main() {
  println!("{}", fib(5));
  println!("{}", fib(12));
  println!("{}", fib(32));
  println!("{}", fib(38));

  println!("{}", memoized_fib(5, &mut HashMap::new()));
  println!("{}", memoized_fib(12, &mut HashMap::new()));
  println!("{}", memoized_fib(32, &mut HashMap::new()));
  println!("{}", memoized_fib(38, &mut HashMap::new()));
  println!("{}", memoized_fib(45, &mut HashMap::new()));
  println!("{}", memoized_fib(110, &mut HashMap::new()));
}